In [50]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [51]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

In [52]:
file_path = "../datos originales/homicidios.xlsx"
df1 = pd.read_excel(file_path, sheet_name="HECHOS")

In [53]:
# Mostrar los registros que contienen guion bajo en alguna columna
df_con_guion_bajo = df1[df1.apply(lambda row: any('_' in str(value) for value in row), axis=1)]

df_con_guion_bajo

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA HH  \
583  2020-0063           2 2020-12-05  2020  12   5  07:10:00  7   

              LUGAR_DEL_HECHO TIPO_DE_CALLE       Calle  Altura        Cruce  \
583  NUEVA YORK Y ALTA GRACIA         CALLE  NUEVA YORK     NaN  ALTA GRACIA   

        Dirección Normalizada  COMUNA                               XY (CABA)  \
583  NUEVA YORK y ALTA GRACIA      11  Point (94080.62190808 102083.62453795)   

            pos x         pos y     PARTICIPANTES      VICTIMA ACUSADO  
583  -58.52783814  -34.61047001  PEATON_MOTO-MOTO  PEATON_MOTO    MOTO

In [54]:
ruta_hechos = "../datos originales/homicidios.xlsx"
df_hechos = pd.read_excel(ruta_hechos, sheet_name="HECHOS")

ruta_victimas = "../datos originales/homicidios.xlsx"
df_victimas = pd.read_excel(ruta_victimas, sheet_name="VICTIMAS")

In [55]:
print(df_hechos.shape)
df_hechos.head()

(696, 21)


ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE             Calle  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  PIEDRA BUENA AV.   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ    PAZ, GRAL. AV.   
2                         AV ENTRE RIOS 2034       AVENIDA    ENTRE RIOS AV.   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA    LARRAZABAL AV.   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA      SAN JUAN AV.   

   Altura                                Cruce  \
0     NaN  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1     NaN                  DE LOS CORRALES AV.   
2  2034.0                                  NaN   
3     NaN             VILLEGAS, CONRADO, GRAL.   
4     NaN              SAENZ PE?A, LUIS, PRES.   

                               Dirección Normalizada  COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...       8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.       9   
2                                ENTRE RIOS AV. 2034       1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.       8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.       1   

                                 XY (CABA)         pos x         pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  
3         MOTO-SD    MOTO         SD  
4  MOTO-PASAJEROS    MOTO  PASAJEROS

In [56]:
df_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Altura                 129 non-null    float64       
 12  Cruce                  525 non-null    object        
 13  Direc

In [57]:
print(df_victimas.shape)
df_victimas.head()

(717, 10)


ID_hecho      FECHA  AAAA  MM  DD        ROL VICTIMA       SEXO EDAD  \
0  2016-0001 2016-01-01  2016   1   1  CONDUCTOR    MOTO  MASCULINO   19   
1  2016-0002 2016-01-02  2016   1   2  CONDUCTOR    AUTO  MASCULINO   70   
2  2016-0003 2016-01-03  2016   1   3  CONDUCTOR    MOTO  MASCULINO   30   
3  2016-0004 2016-01-10  2016   1  10  CONDUCTOR    MOTO  MASCULINO   18   
4  2016-0005 2016-01-21  2016   1  21  CONDUCTOR    MOTO  MASCULINO   29   

   FECHA_FALLECIMIENTO  
0  2016-01-01 00:00:00  
1  2016-01-02 00:00:00  
2  2016-01-03 00:00:00  
3                   SD  
4  2016-02-01 00:00:00

In [58]:
df_victimas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID_hecho             717 non-null    object        
 1   FECHA                717 non-null    datetime64[ns]
 2   AAAA                 717 non-null    int64         
 3   MM                   717 non-null    int64         
 4   DD                   717 non-null    int64         
 5   ROL                  717 non-null    object        
 6   VICTIMA              717 non-null    object        
 7   SEXO                 717 non-null    object        
 8   EDAD                 717 non-null    object        
 9   FECHA_FALLECIMIENTO  717 non-null    object        
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 56.1+ KB


El dataframe de hechos tiene 696 entradas (filas) y 21 columnas.

El dataframe de victimas tiene 717 entradas (filas) y 10 columnas.

Observamos que se pueden unir ambos dataframes mediante el ID del hecho, pero antes de eso , vamos a eliminar las columnas del dataframe de victimas que ya esten presentes en el dataframe de hechos, esto lo haremos dejando solo presentes las no repetidas.

In [59]:
victimas_columnas_repetidas = ['ID_hecho', 'ROL', 'SEXO','EDAD']

# Eliminar las columnas especificadas
df_victimas = df_victimas[victimas_columnas_repetidas]

In [60]:
# Unir dataframes mediante el ID del hecho
df = pd.merge(df_hechos, df_victimas, left_on='ID', right_on='ID_hecho', how='left')

df = df.drop("ID_hecho", axis=1)

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     717 non-null    object        
 1   N_VICTIMAS             717 non-null    int64         
 2   FECHA                  717 non-null    datetime64[ns]
 3   AAAA                   717 non-null    int64         
 4   MM                     717 non-null    int64         
 5   DD                     717 non-null    int64         
 6   HORA                   717 non-null    object        
 7   HH                     717 non-null    object        
 8   LUGAR_DEL_HECHO        717 non-null    object        
 9   TIPO_DE_CALLE          717 non-null    object        
 10  Calle                  716 non-null    object        
 11  Altura                 133 non-null    float64       
 12  Cruce                  540 non-null    object        
 13  Direc

La primera intervencion sera estandarizar la escritura de todos los valores del set de datos a minúsculas para obtener consistencia, facilidad en busquedas y comparaciones. Los encabezados de las columnas se modificaran para que esten escritos completamente en mayusculas.

In [62]:
# Encabezados en mayúsculas
df.columns = df.columns.str.upper()

# Seleccionar solo las columnas no numéricas
columnas_no_numericas = df.select_dtypes(exclude=['number']).columns

# Aplicar la conversión a minúsculas en esas columnas
df[columnas_no_numericas] = df[columnas_no_numericas].map(lambda x: x.lower() if isinstance(x, str) else x)


Realizaremos el proceso de limpieza teniendo en cuenta las situaciones más comunes:

1. Datos faltantes en algunas celdas
2. Columnas irrelevantes (que no responden al problema que queremos resolver)
3. Registros (filas) repetidos
4. Valores extremos (*outliers*) en el caso de las variables numéricas.

Al final de este proceso de limpieza deberíamos tener un set de datos **íntegro**, listo para la fase de Análisis Exploratorio.

In [63]:
print(df.shape)
df.head()

(717, 24)


ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE             CALLE  \
0  av piedra buena y av fernandez de la cruz       avenida  piedra buena av.   
1          av gral paz  y av de los corrales      gral paz    paz, gral. av.   
2                         av entre rios 2034       avenida    entre rios av.   
3      av larrazabal y gral villegas conrado       avenida    larrazabal av.   
4  av san juan y presidente luis saenz peña        avenida      san juan av.   

   ALTURA                                CRUCE  \
0     NaN  fernandez de la cruz, f., gral. av.   
1     NaN                  de los corrales av.   
2  2034.0                                  NaN   
3     NaN             villegas, conrado, gral.   
4     NaN              saenz pe?a, luis, pres.   

                               DIRECCIÓN NORMALIZADA  COMUNA  \
0  piedra buena av. y fernandez de la cruz, f., g...       8   
1               paz, gral. av. y de los corrales av.       9   
2                                entre rios av. 2034       1   
3          larrazabal av. y villegas, conrado, gral.       8   
4            san juan av. y saenz peã‘a, luis, pres.       1   

                                 XY (CABA)         POS X         POS Y  \
0    point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    PARTICIPANTES VICTIMA    ACUSADO        ROL       SEXO EDAD  
0       moto-auto    moto       auto  conductor  masculino   19  
1  auto-pasajeros    auto  pasajeros  conductor  masculino   70  
2       moto-auto    moto       auto  conductor  masculino   30  
3         moto-sd    moto         sd  conductor  masculino   18  
4  moto-pasajeros    moto  pasajeros  conductor  masculino   29

In [64]:
duplicados_totales = df.duplicated().sum()
duplicados_totales

0

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     717 non-null    object        
 1   N_VICTIMAS             717 non-null    int64         
 2   FECHA                  717 non-null    datetime64[ns]
 3   AAAA                   717 non-null    int64         
 4   MM                     717 non-null    int64         
 5   DD                     717 non-null    int64         
 6   HORA                   717 non-null    object        
 7   HH                     717 non-null    object        
 8   LUGAR_DEL_HECHO        717 non-null    object        
 9   TIPO_DE_CALLE          717 non-null    object        
 10  CALLE                  716 non-null    object        
 11  ALTURA                 133 non-null    float64       
 12  CRUCE                  540 non-null    object        
 13  DIREC

Con el fin de evitar redundancia en los datos, eliminaremos las columnas que tengan la fecha y la hora por separado, estaso son "AAAA","MM","DD","HH". La información contenida en estas columnas ya se encuentre en "FECHA" y "HORA"

In [66]:
# Lista de columnas que quieres eliminar
columnas_a_eliminar = ["AAAA", "MM", "DD", "HH"]

# Elimina las columnas especificadas
df = df.drop(columnas_a_eliminar, axis=1)

Los datos no están completos, pues no todas las columnas tienen la misma cantidad de registros.

El número total de registros debería ser 717. Sin embargo las columnas "Altura", "Cruce", "Dirección Normalizada" y "Calle" tienen datos faltantes.

Por ser un set de datos reducido, optaremos por imputar las filas correspondientes.

Revisaremos en primer lugar los registros faltantes en "Calle" y "Dirección Normalizada" para ver que ocurre en esos casos, considerando que es una cantidad pequeña.

In [67]:
# Filtrar registros donde la columna "Calle" es nula
calle_nula = df[df['CALLE'].isnull()]

# Filtrar registros donde la columna "Direccion normalizada" es nula
direccion_nula = df[df['DIRECCIÓN NORMALIZADA'].isnull()]

In [68]:
calle_nula

ID  N_VICTIMAS      FECHA      HORA LUGAR_DEL_HECHO TIPO_DE_CALLE  \
121  2016-0151           1 2016-11-18  20:35:00              sd         calle   

    CALLE  ALTURA CRUCE DIRECCIÓN NORMALIZADA  COMUNA    XY (CABA) POS X  \
121   NaN     NaN   NaN                   NaN       0  point (. .)     .   

    POS Y PARTICIPANTES VICTIMA ACUSADO     ROL SEXO EDAD  
121     .     peaton-sd  peaton      sd  peaton   sd   sd

Al ser solo un registro que cuenta con varios campos sin especificar, se decide eliminarlo del dataset

In [69]:
df = df[df['ID'] != '2016-0151']

In [70]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 716 entries, 0 to 716
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     716 non-null    object        
 1   N_VICTIMAS             716 non-null    int64         
 2   FECHA                  716 non-null    datetime64[ns]
 3   HORA                   716 non-null    object        
 4   LUGAR_DEL_HECHO        716 non-null    object        
 5   TIPO_DE_CALLE          716 non-null    object        
 6   CALLE                  716 non-null    object        
 7   ALTURA                 133 non-null    float64       
 8   CRUCE                  540 non-null    object        
 9   DIRECCIÓN NORMALIZADA  708 non-null    object        
 10  COMUNA                 716 non-null    int64         
 11  XY (CABA)              716 non-null    object        
 12  POS X                  716 non-null    object        
 13  POS Y     

In [71]:
direccion_nula

ID  N_VICTIMAS      FECHA      HORA  \
39   2016-0052           1 2016-04-20  20:00:00   
108  2016-0136           1 2016-10-25  00:00:00   
121  2016-0151           1 2016-11-18  20:35:00   
186  2017-0050           2 2017-04-28  11:08:08   
187  2017-0050           2 2017-04-28  11:08:08   
188  2017-0051           1 2017-05-01  03:47:47   
327  2018-0039           1 2018-04-21  22:15:00   
564  2020-0026           1 2020-05-17  06:40:00   
642  2021-0023           1 2021-03-01  09:20:00   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  \
39                autopista lugones pk 10000     autopista   
108         au buenos aires - la plata km. 4     autopista   
121                                       sd         calle   
186  au perito moreno y ramal enlace au1/au6     autopista   
187  au perito moreno y ramal enlace au1/au6     autopista   
188                       au dellepiane 2400     autopista   
327                 autopista lugones km 4.7     autopista   
564             lugones, leopoldo av. km 6,1     autopista   
642         au buenos aires la plata  km 4,5     autopista   

                                    CALLE  ALTURA CRUCE DIRECCIÓN NORMALIZADA  \
39                  lugones, leopoldo av.     NaN   NaN                   NaN   
108     autopista buenos aires - la plata     NaN   NaN                   NaN   
121                                   NaN     NaN   NaN                   NaN   
186               autopista perito moreno     NaN   NaN                   NaN   
187               autopista perito moreno     NaN   NaN                   NaN   
188  autopista dellepiane luis tte. gral.     NaN   NaN                   NaN   
327                 lugones, leopoldo av.     NaN   NaN                   NaN   
564                 lugones, leopoldo av.     NaN   NaN                   NaN   
642     autopista buenos aires - la plata     NaN   NaN                   NaN   

     COMUNA    XY (CABA) POS X POS Y     PARTICIPANTES VICTIMA      ACUSADO  \
39       13  point (. .)     .     .           moto-sd    moto           sd   
108       4  point (. .)     .     .       moto-cargas    moto       cargas   
121       0  point (. .)     .     .         peaton-sd  peaton           sd   
186       9  point (. .)     .     .       moto-cargas    moto       cargas   
187       9  point (. .)     .     .       moto-cargas    moto       cargas   
188       7  point (. .)     .     .         auto-auto    auto         auto   
327      14  point (. .)     .     .       peaton-auto  peaton         auto   
564      14  point (. .)     .     .  moto-objeto fijo    moto  objeto fijo   
642       4  point (. .)     .     .       moto-cargas    moto       cargas   

                      ROL       SEXO EDAD  
39                     sd         sd   sd  
108             conductor         sd   sd  
121                peaton         sd   sd  
186             conductor  masculino   46  
187  pasajero_acompañante  masculino   16  
188             conductor  masculino   33  
327                peaton  masculino   37  
564             conductor  masculino   28  
642             conductor  masculino   24

Notamos que todos los datos asociados a la ubicación no estan presentes en estos registros, por lo que se decide imputar como "sin datos", y tomar la columna Altura como un string ya que no tiene un sentido numérico continuo.

In [72]:
# Convertir la columna 'Altura' a tipo de dato string
# Convertir la columna 'Altura' a tipo de dato string y reemplazar NaN por 'None'
df['ALTURA'] = df['ALTURA'].astype(str).replace('nan', None)

# Imputación para Altura en registros con Dirección Normalizada nula
df.loc[df['DIRECCIÓN NORMALIZADA'].isnull(), 'ALTURA'] = 'sin datos'

# Imputación para Cruce en registros con Dirección Normalizada nula
df.loc[df['DIRECCIÓN NORMALIZADA'].isnull(), 'CRUCE'] = 'sin datos'

# Imputación para XY (CABA) en registros con Dirección Normalizada nula
df.loc[df['DIRECCIÓN NORMALIZADA'].isnull(), 'XY (CABA)'] = 'sin datos'

# Imputación para pos x y pos y en registros con Dirección Normalizada nula
df.loc[df['DIRECCIÓN NORMALIZADA'].isnull(), 'POS X'] = 'sin datos'
df.loc[df['DIRECCIÓN NORMALIZADA'].isnull(), 'POS Y'] = 'sin datos'

# Imputación para Dirección Normalizada
df['DIRECCIÓN NORMALIZADA'].fillna('sin datos', inplace=True)

Se verifica la ausencia de registros con datos asociados a la direccion nulos, ya que fueron imputados.

In [73]:
direccion_nula = df[df['DIRECCIÓN NORMALIZADA'].isnull()]
direccion_nula

Empty DataFrame
Columns: [ID, N_VICTIMAS, FECHA, HORA, LUGAR_DEL_HECHO, TIPO_DE_CALLE, CALLE, ALTURA, CRUCE, DIRECCIÓN NORMALIZADA, COMUNA, XY (CABA), POS X, POS Y, PARTICIPANTES, VICTIMA, ACUSADO, ROL, SEXO, EDAD]
Index: []

In [74]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 716 entries, 0 to 716
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     716 non-null    object        
 1   N_VICTIMAS             716 non-null    int64         
 2   FECHA                  716 non-null    datetime64[ns]
 3   HORA                   716 non-null    object        
 4   LUGAR_DEL_HECHO        716 non-null    object        
 5   TIPO_DE_CALLE          716 non-null    object        
 6   CALLE                  716 non-null    object        
 7   ALTURA                 141 non-null    object        
 8   CRUCE                  548 non-null    object        
 9   DIRECCIÓN NORMALIZADA  716 non-null    object        
 10  COMUNA                 716 non-null    int64         
 11  XY (CABA)              716 non-null    object        
 12  POS X                  716 non-null    object        
 13  POS Y     

Ahora nos falta revisar un poco mas como tratar las columnas "Altura" y "Cruce". Al sumar las cantidades de no nulos en estas dos columnas (136 + 532 = 668) nos permite pensar que podria estar ocurriendo que para cuando una "Altura" no esta definida, es porque el hecho fue en un "Cruce" de calles y viceversa. Sin embargo revisaremos los valores donde no estan definidas "Cruce" y "Altura" simultaneamente.

In [75]:
cruce_altura_nulos = df.loc[(df['CRUCE'].isnull()) & (df['ALTURA'].isnull())]
cruce_altura_nulos

ID  N_VICTIMAS      FECHA      HORA  \
36   2016-0049           1 2016-04-17  00:00:00   
65   2016-0087           1 2016-07-02  00:10:00   
72   2016-0096           1 2016-07-25  07:00:00   
82   2016-0107           1 2016-08-20  08:22:00   
92   2016-0118           1 2016-09-04  03:30:00   
102  2016-0130           1 2016-10-04  12:30:00   
129  2016-0160           1 2016-12-06  05:30:00   
141  2016-0174           1 2016-12-27  00:00:00   
150  2017-0009           1 2017-01-16  13:56:00   
157  2017-0016           1 2017-02-03  05:12:21   
182  2017-0042           1 2017-04-10  09:00:00   
242  2017-0112           2 2017-09-10  09:32:17   
243  2017-0112           2 2017-09-10  09:32:17   
266  2017-0140           1 2017-11-19  23:22:17   
292  2018-0007           1 2018-01-11  02:30:00   
306  2018-0020           1 2018-02-19  22:00:00   
360  2018-0072           1 2018-07-11  13:15:00   
401  2018-0111           1 2018-10-19  12:13:00   
414  2018-0124           1 2018-11-15  19:40:00   
492  2019-0058           1 2019-07-10  09:20:00   
510  2019-0076           1 2019-09-05  18:00:00   
529  2019-0095           1 2019-11-20  15:30:00   
531  2019-0097           1 2019-11-25  08:50:00   
573  2020-0035           1 2020-08-04  05:30:00   
578  2020-0039           1 2020-09-01  19:17:42   
605  2020-0066           1 2020-12-07  04:30:00   
620  2021-0001           1 2021-01-06  18:00:00   
622  2021-0003           1 2021-01-10  07:40:00   
645  2021-0026           1 2021-03-15  02:35:00   
652  2021-0033           1 2021-04-14  17:50:00   
664  2021-0045           1 2021-05-26  17:24:00   
676  2021-0057           1 2021-07-10  23:30:00   
686  2021-0067           1 2021-08-08  09:29:00   
700  2021-0081           1 2021-11-02  13:35:00   
701  2021-0082           1 2021-11-06  04:39:00   

                                       LUGAR_DEL_HECHO TIPO_DE_CALLE  \
36   autopista 1 sur presidente arturo frondizi km....     autopista   
65    autopista 1 sur pte arturo frondizi y av caseros     autopista   
72        autopista dellepiane luis tte. gral. km. 2,3     autopista   
82             autopista 9 de julio sur alt av mendoza     autopista   
92          autopista 1 sur pte arturo frondizi km 2.9     autopista   
102               autopista 9 de julio sur y av brasil     autopista   
129                autopista perito moreno altura velz     autopista   
141                               autopista 25 de mayo     autopista   
150                          au 25 de mayo y av. boedo     autopista   
157  au 25 de mayo, kilometro 8.3, sentido hacia pr...     autopista   
182                      av. leopoldo lugones pkm 6900      gral paz   
242                         au 25 de mayo y av. varela     autopista   
243                         au 25 de mayo y av. varela     autopista   
266                        au arturo frondizi pkm 3100     autopista   
292                           au perito moreno  km 0.6     autopista   
306   fernandez de la cruz subida au héctor j. cámpora     autopista   
360  au dellepiane kilometro 0.0, altura calle mont...     autopista   
401    au perito moreno y av. eva perón sentido centro     autopista   
414          au perito moreno y reservistas argentinos     autopista   
492  autopista perito moreno altura paso a nivel de...     autopista   
510  autopista 1 sur presidente arturo frondizi alt...     autopista   
529  autopista 1 sur presidente arturo frondizi alt...     autopista   
531                     autopista perito moreno km 5,6     autopista   
573  au 25 de mayo altura pumacahua y recuero, casi...     autopista   
578                                    murguiondo 2700         calle   
605  autopista perito moreno altura estacion de tre...     autopista   
620                                  au frondizi km. 3     autopista   
622                              au 25 de mayo km. 3,7     autopista   
645        au frondizi altura quinquela martin, benito     autopista   
652            

Observamos que los casos en los que la altura y la calle no estan definidas simultaneamente, pero en la mayoria de los casos tenemos informacion de la pos x y pos y por separado. Tambien hay algunos casos donde la posicion no esta definida en ninguna parte. Finalmente optamos por imputar todos los casos por "sin dato" ya que aun sin esa información especificada, los registros contienen otros valores que pueden ser utiles para otros analisis como el nombre y tipo de calle

In [76]:
df['CRUCE'].fillna('sin datos', inplace=True)
df['ALTURA'].fillna('sin datos', inplace=True)

Ya no tenemos nulos en nuestro set de datos

In [77]:
cruce_altura_nulos = df.loc[(df['CRUCE'].isnull()) & (df['ALTURA'].isnull())]
cruce_altura_nulos

Empty DataFrame
Columns: [ID, N_VICTIMAS, FECHA, HORA, LUGAR_DEL_HECHO, TIPO_DE_CALLE, CALLE, ALTURA, CRUCE, DIRECCIÓN NORMALIZADA, COMUNA, XY (CABA), POS X, POS Y, PARTICIPANTES, VICTIMA, ACUSADO, ROL, SEXO, EDAD]
Index: []

In [78]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 716 entries, 0 to 716
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     716 non-null    object        
 1   N_VICTIMAS             716 non-null    int64         
 2   FECHA                  716 non-null    datetime64[ns]
 3   HORA                   716 non-null    object        
 4   LUGAR_DEL_HECHO        716 non-null    object        
 5   TIPO_DE_CALLE          716 non-null    object        
 6   CALLE                  716 non-null    object        
 7   ALTURA                 716 non-null    object        
 8   CRUCE                  716 non-null    object        
 9   DIRECCIÓN NORMALIZADA  716 non-null    object        
 10  COMUNA                 716 non-null    int64         
 11  XY (CABA)              716 non-null    object        
 12  POS X                  716 non-null    object        
 13  POS Y     

En la columna "Participantes" ocurre algo particular. Estan los datos de dos participantes separados por guión medio. Esto pudimos verlo en impresiones del dataframe *cruce_altura_nulos*. Para mayor comodidad y poder utilizar esta informacion mas facilmente en gráficos, separaremos estos datos en 2 columnas. "Participante A" y "Participante B"

In [79]:
# Dividir la columna 'Participantes' en dos nuevas columnas
df[['PARTICIPANTE A', 'PARTICIPANTE B']] = df['PARTICIPANTES'].str.split('-', n=1, expand=True)

# Eliminar la columna 'Participantes'
df.drop('PARTICIPANTES', axis=1, inplace=True)

In [80]:
df.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'HORA', 'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE',
       'CALLE', 'ALTURA', 'CRUCE', 'DIRECCIÓN NORMALIZADA', 'COMUNA',
       'XY (CABA)', 'POS X', 'POS Y', 'VICTIMA', 'ACUSADO', 'ROL', 'SEXO',
       'EDAD', 'PARTICIPANTE A', 'PARTICIPANTE B'],
      dtype='object')

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 716 entries, 0 to 716
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     716 non-null    object        
 1   N_VICTIMAS             716 non-null    int64         
 2   FECHA                  716 non-null    datetime64[ns]
 3   HORA                   716 non-null    object        
 4   LUGAR_DEL_HECHO        716 non-null    object        
 5   TIPO_DE_CALLE          716 non-null    object        
 6   CALLE                  716 non-null    object        
 7   ALTURA                 716 non-null    object        
 8   CRUCE                  716 non-null    object        
 9   DIRECCIÓN NORMALIZADA  716 non-null    object        
 10  COMUNA                 716 non-null    int64         
 11  XY (CABA)              716 non-null    object        
 12  POS X                  716 non-null    object        
 13  POS Y     

Imputamos nulos en la nueva columna PARTICIPANTE B con "sin dato"

In [82]:
df['PARTICIPANTE B'].fillna("sin datos", inplace=True)

Ya no tenemos nulos en nuestro set de datos y dividimos la columna "Participantes", pero aun queda estandarizar las imputaciones de valores, ya que por defecto, cuando no hay datos viene rellenado como "sd".

In [83]:
df.replace('sd', 'sin datos', inplace=True)

Se observó que la columna "LUGAR_DEL_HECHO" contiene una combinacion de la información de las columnas "Altura", "Cruce", "Calle" y en ocasiones alguna información extra no muy relevante que incluye numeros y "km". Se decide eliminar esta columna ya que para los fines del proyecto se desean los datos separados para poder evaluar relaciones y hacer interpretaciones.

In [84]:
df = df.drop('LUGAR_DEL_HECHO', axis=1)

Ahora revisaremos las 3 columnas dedicadas a posición, "XY (CABA)" y "pos x" / "pos y". En el primer caso tenemos geocodificacion plana, y en el segundo (las columnas de pos x y pos y) tenemos el sistema wgs84. Se verá cual tiene mayor cantidad de informacion, y si ocurre que una tiene informacion y la otra no. 

Primero imputaremos los datos con "sin dato". Notamos que cuando no hay informacion, en XY (CABA) se ve "point(. .)", mientras que en las otras dos columnas se rellenan con "." puntos. 

In [85]:
df['XY (CABA)'] = df['XY (CABA)'].replace('point (. .)', 'sin datos')
df['POS X'] = df['POS X'].replace('.', 'sin datos')
df['POS Y'] = df['POS Y'].replace('.', 'sin datos')

In [86]:
# Contar cuando 'XY (CABA)' tiene datos y 'pos x' y 'pos y' están ambos sin datos
conteo_datos_faltantes = df.apply(lambda x: ('sin datos' in x['POS X'] and 'sin datos' in x['POS Y'] and 'sin datos' not in x['XY (CABA)']), axis=1).value_counts()

In [87]:
conteo_datos_faltantes

False    716
Name: count, dtype: int64

In [88]:
# Contar cuando 'pos x' y 'pos y' tienen datos y 'XY (CABA)' está sin datos
conteo_datos_presentes = df.apply(lambda x: ('sin datos' in x['XY (CABA)'] and 'sin datos' not in x['POS X'] and 'sin datos' not in x['POS Y']), axis=1).value_counts()

In [89]:
conteo_datos_presentes

False    714
True       2
Name: count, dtype: int64

Como observamos, hay 2 casos en los que la columna XY (CABA) no tiene datos, pero la pos x y pos y si tienen. Se decide eliminar la primer columna y trabajar entonces con las columnas en geocodificacion wgs84

In [90]:
# Eliminar la columna 'XY (CABA)'
df = df.drop(columns=['XY (CABA)'])

In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 716 entries, 0 to 716
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     716 non-null    object        
 1   N_VICTIMAS             716 non-null    int64         
 2   FECHA                  716 non-null    datetime64[ns]
 3   HORA                   716 non-null    object        
 4   TIPO_DE_CALLE          716 non-null    object        
 5   CALLE                  716 non-null    object        
 6   ALTURA                 716 non-null    object        
 7   CRUCE                  716 non-null    object        
 8   DIRECCIÓN NORMALIZADA  716 non-null    object        
 9   COMUNA                 716 non-null    int64         
 10  POS X                  716 non-null    object        
 11  POS Y                  716 non-null    object        
 12  VICTIMA                716 non-null    object        
 13  ACUSADO   

Crearemos la variable categórica `rango_etario` a partir de la variable `EDAD` con rangos etarios comprendidos entre `0-17`, `18-29`, `30-44`, `45-64` y `65+`, con el objetivo de facilitar el análisis exploratorio de datos. Para esto, procedemos de la siguiente forma:

In [92]:
# Convertimos la columna 'EDAD' a tipo de dato numérico
df['EDAD'] = pd.to_numeric(df['EDAD'], errors='coerce')

# Definimos los límites de los intervalos de edad
bins = [0, 18, 30, 45, 65, 100]

# Definimos las etiquetas de los intervalos de edad
labels = ['0-17', '18-29', '30-44', '45-64', '65+']

# Creamos una nueva columna 'rango_edad' con los intervalos de edad
df['RANGO_ETARIO'] = pd.cut(df['EDAD'], bins=bins, labels=labels, right=False)

# Agregamos la categoría 'SD' a la columna 'rango_edad'
df['RANGO_ETARIO'] = df['RANGO_ETARIO'].cat.add_categories('sin datos')
df[['EDAD','RANGO_ETARIO']] = df[['EDAD','RANGO_ETARIO']].fillna('sin datos')

# Verificamos el resultado
df[['EDAD','RANGO_ETARIO']].sample(5)

EDAD RANGO_ETARIO
607  71.0          65+
512  72.0          65+
303  35.0        30-44
381  23.0        18-29
424  60.0        45-64

Crearemos una nueva columna llamada "ENCRUCIJADA" donde asignaremos un valor booleando dependiendo de si el siniesto fue o no en un cruce.

In [93]:
# Crear la nueva columna "ENCRUCIJADA" con valores iniciales en False
df['ENCRUCIJADA'] = False

# Actualizar los valores a True donde la columna 'CRUCE' no sea 'sin dato'
df.loc[df['CRUCE'] != 'sin datos', 'ENCRUCIJADA'] = True

Imputamos el unico valor faltante en HORA con la media

In [94]:
# Reemplazar el valor "sin datos" con el valor del registro de abajo
df['HORA'].replace('sin datos', method='ffill', inplace=True)


C:\Users\maupe\AppData\Local\Temp\ipykernel_20828\3538326915.py:2: FutureWarning: The 'method' keyword in Series.replace is deprecated and will be removed in a future version.
  df['HORA'].replace('sin datos', method='ffill', inplace=True)


Nos aseguramos de que la columna FECHA tenga un formato date y la columna HORA un formato time.

In [95]:
df['FECHA'] = pd.to_datetime(df['FECHA'].dt.strftime('%Y-%m-%d'))

# Convertir la columna 'hora' a tipo string para manipularla
df['HORA'] = df['HORA'].astype(str)

# Identificar y extraer solo la hora
df['HORA'] = df['HORA'].apply(lambda x: x.split()[-1])
df['HORA'] = pd.to_datetime(df['HORA'], format='%H:%M:%S').dt.time

Ahora todas las fechas tienen un formato unificado.

In [96]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 716 entries, 0 to 716
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     716 non-null    object        
 1   N_VICTIMAS             716 non-null    int64         
 2   FECHA                  716 non-null    datetime64[ns]
 3   HORA                   716 non-null    object        
 4   TIPO_DE_CALLE          716 non-null    object        
 5   CALLE                  716 non-null    object        
 6   ALTURA                 716 non-null    object        
 7   CRUCE                  716 non-null    object        
 8   DIRECCIÓN NORMALIZADA  716 non-null    object        
 9   COMUNA                 716 non-null    int64         
 10  POS X                  716 non-null    object        
 11  POS Y                  716 non-null    object        
 12  VICTIMA                716 non-null    object        
 13  ACUSADO   

Luego de empezar a revisar algunos comportamientos en el EDA, se notó que la columna VICTIMA, contiene algunos valores extraños como "peaton_moto" y "objeto fijo", los cuales no estan en el diccionario asociado a los posibles valores de la columna.

Se procede a inspeccionar estos casos:

In [97]:
df[df['VICTIMA'].isin(['objeto fijo', 'peaton_moto'])]

ID  N_VICTIMAS      FECHA      HORA TIPO_DE_CALLE           CALLE  \
237  2017-0108           2 2017-09-02  04:53:08      gral paz  paz, gral. av.   
238  2017-0108           2 2017-09-02  04:53:08      gral paz  paz, gral. av.   
602  2020-0063           2 2020-12-05  07:10:00         calle      nueva york   
603  2020-0063           2 2020-12-05  07:10:00         calle      nueva york   

        ALTURA        CRUCE     DIRECCIÓN NORMALIZADA  COMUNA         POS X  \
237  sin datos      machain  paz, gral. av. y machain      12  -58.49491054   
238  sin datos      machain  paz, gral. av. y machain      12  -58.49491054   
602  sin datos  alta gracia  nueva york y alta gracia      11  -58.52783814   
603  sin datos  alta gracia  nueva york y alta gracia      11  -58.52783814   

            POS Y      VICTIMA ACUSADO                   ROL       SEXO  EDAD  \
237  -34.54795581  objeto fijo    auto  pasajero_acompañante  masculino  17.0   
238  -34.54795581  objeto fijo    auto  pasajero_acompañante  masculino  10.0   
602  -34.61047001  peaton_moto    moto                peaton  masculino  38.0   
603  -34.61047001  peaton_moto    moto  pasajero_acompañante  masculino  21.0   

    PARTICIPANTE A PARTICIPANTE B RANGO_ETARIO  ENCRUCIJADA  
237           auto    objeto fijo         0-17         True  
238           auto    objeto fijo         0-17         True  
602    peaton_moto           moto        30-44         True  
603    peaton_moto           moto        18-29         True

Podemos ver que el ACUSADO y la VICTIMA tienen valores que podrían estar invertidos segundo el diccionario de datos, de modo que basandonos en el mismo, hacemos un intercambio de estos. Tambien se puede ver como en las ultimas dos filas, el ACUSADO es "moto", por lo que corregiremos la victima a solo "peaton", asumiento que la informacion es correcta a pesar de estar invertida.

In [98]:
# Invertimos los valores de las columnas 'VICTIMA' y 'ACUSADO' en los índices 237 y 238
df.loc[[237, 238], ['VICTIMA', 'ACUSADO']] = df.loc[[237, 238], ['ACUSADO', 'VICTIMA']].values

# Reemplazamos los valores de las columnas 'PARTICIPANTES' y 'VICTIMA' en los índices 602 y 603
df.loc[[602, 603], ['PARTICIPANTE A', 'VICTIMA']] = df.loc[[602, 603], ['PARTICIPANTE A', 'VICTIMA']].replace({'peaton_moto': 'peaton-moto', 'peaton_moto': 'peaton'})
registros_seleccionados = df[df['ID'].isin(['2017-0108', '2020-0063'])]

In [99]:
registros_seleccionados

ID  N_VICTIMAS      FECHA      HORA TIPO_DE_CALLE           CALLE  \
237  2017-0108           2 2017-09-02  04:53:08      gral paz  paz, gral. av.   
238  2017-0108           2 2017-09-02  04:53:08      gral paz  paz, gral. av.   
602  2020-0063           2 2020-12-05  07:10:00         calle      nueva york   
603  2020-0063           2 2020-12-05  07:10:00         calle      nueva york   

        ALTURA        CRUCE     DIRECCIÓN NORMALIZADA  COMUNA         POS X  \
237  sin datos      machain  paz, gral. av. y machain      12  -58.49491054   
238  sin datos      machain  paz, gral. av. y machain      12  -58.49491054   
602  sin datos  alta gracia  nueva york y alta gracia      11  -58.52783814   
603  sin datos  alta gracia  nueva york y alta gracia      11  -58.52783814   

            POS Y VICTIMA      ACUSADO                   ROL       SEXO  EDAD  \
237  -34.54795581    auto  objeto fijo  pasajero_acompañante  masculino  17.0   
238  -34.54795581    auto  objeto fijo  pasajero_acompañante  masculino  10.0   
602  -34.61047001  peaton         moto                peaton  masculino  38.0   
603  -34.61047001  peaton         moto  pasajero_acompañante  masculino  21.0   

    PARTICIPANTE A PARTICIPANTE B RANGO_ETARIO  ENCRUCIJADA  
237           auto    objeto fijo         0-17         True  
238           auto    objeto fijo         0-17         True  
602         peaton           moto        30-44         True  
603         peaton           moto        18-29         True

Finalmente tenemos nuestro set de datos limpio, sin nulos ni duplicados. Los valores faltantes fueron imputados y se han eliminado columnas innecesarias.

In [100]:
df.to_csv('../datos procesados/homicidios_clean.csv', index=False,date_format='%Y-%m-%d')